syntax = "proto3";

package raft;

// =======================
// RAFT SERVICE
// =======================
service RaftService {

  // Leader election
  rpc RequestVote (RequestVoteRequest)
      returns (RequestVoteResponse);

  // Heartbeat + log replication
  rpc AppendEntries (AppendEntriesRequest)
      returns (AppendEntriesResponse);

  // Simulate network partition / disconnect
  rpc Disconnect (DisconnectRequest)
      returns (DisconnectResponse);
}

// =======================
// REQUEST VOTE
// =======================
message RequestVoteRequest {
  int32 term = 1;              
  int32 candidate_id = 2;      
  int32 last_log_index = 3;    
  int32 last_log_term = 4;     
}

message RequestVoteResponse {
  int32 term = 1;              
  bool vote_granted = 2;       
}

// =======================
// APPEND ENTRIES
// =======================
message LogEntry {
  int32 term = 1;
  string command = 2;          
}

message AppendEntriesRequest {
  int32 term = 1;              
  int32 leader_id = 2;
  int32 prev_log_index = 3;    
  int32 prev_log_term = 4;     
  repeated LogEntry entries = 5;  
  int32 leader_commit = 6;       
}

message AppendEntriesResponse {
  int32 term = 1;
  bool success = 2;            
}

// =======================
// NETWORK PARTITION (SIMULATION)
// =======================
message DisconnectRequest {
  repeated int32 node_ids = 1; 
}

message DisconnectResponse {
  bool success = 1;
}
